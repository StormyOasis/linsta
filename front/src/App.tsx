import React, { useEffect, useState } from "react"
import styled from 'styled-components';
import { Provider } from 'react-redux';
import store from './store';

const Title = styled.h1`
  color: red;
`

const App = () => {
 /*8 useEffect(() => {
    fetch("http://localhost:3001")
      .then((response) => response.json())
      .catch((error) => console.log(error))
  }, [])*/

  return (
    <Provider store={store}>
      <div className="container">
        <Title>Linsta - Front</Title>
      </div>
    </Provider>
  )
}

export default App