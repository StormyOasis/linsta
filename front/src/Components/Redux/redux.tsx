import { configureStore} from '@reduxjs/toolkit';

import authSliceCreator from './slices/auth.slice';

export type Actions = {
  authActions?: any,
}

export const actions: Actions = {};

export const buildStore = (initialState?: any) => {
  const preloadedState:any = initialState == null ? {} : initialState;
  const {authActions, authReducer} = authSliceCreator(preloadedState.auth);

  actions.authActions = authActions;

  const store = configureStore({
    reducer: {
      auth: authReducer,
    },
    ...preloadedState
  });

  return store;
}

export type ConfiguredStore = ReturnType<typeof buildStore>;
export type StoreGetState = ConfiguredStore["getState"];
export type RootState = ReturnType<StoreGetState>;
export type AppDispatch = ConfiguredStore["dispatch"];