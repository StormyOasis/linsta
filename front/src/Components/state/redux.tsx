import { configureStore } from "@reduxjs/toolkit";
import rootReducer from "./reducers/rootReducer";
import { initialState } from "./store";

const store = configureStore({
    reducer: rootReducer,
    devTools: true,
    preloadedState: initialState,
  });
  
  export type AppDispatch = typeof store.dispatch;
  export type RootState = ReturnType<typeof store.getState>;
  
  export default store;
  