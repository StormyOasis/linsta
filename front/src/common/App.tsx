import React from "react"
import { Provider } from 'react-redux';
import { ErrorBoundary } from "react-error-boundary";

import Header from "./Components/Header";
import Html from "./render";

const App = (props: any) => {
  return (
    <React.StrictMode>
      <Html title="Linstagram" initialState={props.initialState}>
        <Provider store={props.store}>
          <ErrorBoundary FallbackComponent={({error}) => <div>{error.stack}</div>}>
            <div className="container">            
              <Header />
            </div>
          </ErrorBoundary>
        </Provider>
      </Html>
    </React.StrictMode>
  )
}

export default App;