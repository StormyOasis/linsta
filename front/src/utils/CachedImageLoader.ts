import { clear, del, get, set } from 'idb-keyval';
import { blobToBase64 } from "./utils";

export const loadImageCached = async (id: string, url: string): Promise<string> => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // Image not in the cache so have to fetch it
        try {
            const result = await fetch(url);
            if(result.status === 200) {
                const blob = await result.blob();
                const b64:string = await blobToBase64(blob) as string;

                // add the base64 data into the cache and return it
                await set(id, b64);

                return b64;
            }
        } catch(err) {
            console.log(err);
            return url; //default to what was passed in
        }

    } else {
        // Image is in the cache already as a base64 string, use that as is
        return value;
    }

    return url; //default to what was passed in
}

export const setImage = async (id: string, data: string) => {     
    await set(id, data);
}

export const deleteFromCache = async (id: string) => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // key not present in cache so do nothing
        return;
    }

    // delete from the cache
    await del(id);
}

export const clearCache = async () => {
    await clear();
}