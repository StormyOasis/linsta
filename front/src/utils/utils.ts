import sanitizeHtml from 'sanitize-html';

import { HistoryType, Post, Profile } from "../api/types";
import { postGetProfileByUserId, postSetFollowStatus } from '../api/ServiceController';
import { DEFAULT_PFP } from '../api/config';
import { getProfileByUserId } from 'src/Components/Redux/slices/profile.slice';

export const historyUtils: HistoryType = {
    navigate: null,
    location: null,
    isServer: true
}

export const validatePassword = (value: string): boolean => {
    const regex: RegExp =
        /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@.#$!%*?&^])[A-Za-z\d@.#$!%*?&]{8,15}$/;

    return regex.test(value);
};

export const validateEmailPhone = (value: string): boolean => {
    if (value == null) {
        return false;
    }

    const phoneRegex = /(?:([+]\d{1,4})[-.\s]?)?(?:[(](\d{1,3})[)][-.\s]?)?(\d{1,4})[-.\s]?(\d{1,4})[-.\s]?(\d{1,9})/g;
    const emailRegex =
        /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
    return emailRegex.test(value) || phoneRegex.test(value);
};

export const validateFullName = (value: string): boolean => {
    return (
        value !== null &&
        value.trim().length > 0 &&
        value.trim().split(" ").length > 1
    );
};

export const validateUrl = (text: string|undefined): boolean => {
    if(text == null || text.length === 0) {
        return true; //No url entered should be considered valid
    }

    try {
        new URL(text);
        return true;
    } catch (e) {
        return false;
    }
}

export const isVideoFileFromPath = (path: string): boolean => {
    if (path == null || path.trim().length < 4) {
        throw new Error("Invalid path");
    }

    const validVideoExtentions: string[] = ["avif", "ogm", "wmv", "mpg", "webm", "ogv", "mov", "asx", "mpeg", "mp4", "m4v", "avi"];
    const ext = path.toLowerCase().substring(path.lastIndexOf(".") + 1);

    return validVideoExtentions.includes(ext);
}

export const isVideoFileFromType = (type: string): boolean => {
    if (type == null || type.trim().length < 4) {
        throw new Error("Invalid type");
    }

    return type.toLowerCase().includes("video");
}

export const base64ToBlob = (base64String: string, outFileName: string): File | null => {
    if (base64String === null || base64String.length === 0)
        return null;

    const contentType = base64String.substring(5, base64String.indexOf(';'));
    const data = base64String.substring(base64String.indexOf(",") + 1);

    return base64ToBlobEx(data, contentType, outFileName);
}

export const base64ToBlobEx = (base64String: string, contentType: string, outFileName: string) => {
    const sliceSize = 512;

    var byteCharacters = atob(base64String);
    var byteArrays = [];

    for (var offset = 0; offset < byteCharacters.length; offset += sliceSize) {
        var slice = byteCharacters.slice(offset, offset + sliceSize);

        var byteNumbers = new Array(slice.length);
        for (var i = 0; i < slice.length; i++) {
            byteNumbers[i] = slice.charCodeAt(i);
        }

        var byteArray = new Uint8Array(byteNumbers);

        byteArrays.push(byteArray);
    }

    return new File(byteArrays, outFileName, { type: contentType });
}

export const blobToBase64 = async (blob: any) => {
    blob = await createBlob(blob);

    return new Promise((resolve, _reject) => {
        const reader = new FileReader();
        reader.onloadend = () => resolve(reader.result);
        reader.readAsDataURL(blob);
    });
}

export const createBlob = async (url: string) => {
    try {
        const response = await fetch(url);
        const data = await response.blob();
        return data;
    } catch (err) {
        console.log(err);
    }
    return null;
}

export const extractMimeTypeFromBase64 = (data: string): string | null => {
    if (data == null || data.length === 0) {
        return null;
    }

    return data.substring(5, data.indexOf(';') + 1);
}

export const extractFrameFromVideo = async (video: HTMLVideoElement): Promise<string | null> => {
    if (video == null) {
        return null;
    }

    const canvas = document.createElement("canvas");
    const ctx = canvas.getContext('2d');

    if (ctx == null) {
        return null;
    }

    canvas.height = video.videoHeight || parseInt(video.style.height);
    canvas.width = video.videoWidth || parseInt(video.style.width);

    ctx.drawImage(video, 0, 0);

    return new Promise((resolve, _reject) => {
        canvas.toBlob((file: any) =>
            resolve(URL.createObjectURL(file))
            , 'image/jpeg')
    })
}

export const dateDiff = (dateTime: Date) => {
    if(dateTime == null) {
        return "now";
    }
    
    const now = new Date().getTime();
    const dateTimeMilli = new Date(dateTime).getTime();
    const diff = new Date(now - dateTimeMilli);
    let diffMilli = diff.getTime();

    const days = Math.floor(diffMilli / 1000 / 60 / 60 / 24);
    diffMilli = diffMilli - (days * 1000*60*60*24);

    const weeks = Math.floor(days / 7.0);
    const hours = Math.floor(diffMilli / 1000 / 60 / 60);
    diffMilli = diffMilli - (hours * 1000 * 60 * 60);
    const mins = Math.floor(diffMilli / 1000 / 60);

    if(weeks >= 1) {
        return weeks + "w";
    }

    if(days >= 1) {
        return days + "d";
    }

    if(hours >= 1) {
        return hours + "h";
    } 
    
    if(mins >= 1) {
        return mins + "m";
    }

    return "now";
}

export const getDateAsText = (date: Date) => {
    return new Date(date).toLocaleDateString('en-us', { year: "numeric", month: "long", day: "numeric" });
}

export const getPostFromListById = (postId: string, posts: Post[]):Post => {
    for(let post of posts) {
        if(post.postId === postId) {
            return post;
        }
    }
    return {} as any;
}

export const isPostLiked = (userName: string, post: Post):boolean => {
    if(post == null || userName == null || post.global?.likes == null) {
        return false;
    }

    const result = post.global.likes.filter((like:any) => like.userName === userName);

    return result.length > 0;    
}

export const togglePostLikedState = (userName: string, userId: string, post: Post):(Post|null) => {    
    if(post == null || userName == null || userId == null) {
        return null;
    }
    
    if(post.global.likes == null) {
        post.global.likes = [];
    }

    const index = post.global.likes.findIndex((value:any) => value.userName === userName);
    if(index === -1) {
        // Username is not in the post's like list, so add it
        post.global.likes.push({userName, userId});
    } else {
        // Remove the username from the post's like list
        post.global.likes.splice(index, 1);
    }

    return post;
}

export const getSanitizedText = (text: string):[string, string] => {
    if(text == null || text.length === 0)
        return [text, text];

    const textHtml = sanitizeHtml(text, {
        allowedTags: [ 'b', 'i', 'em', 'strong', 'a', 'br', 'sub', 'sup' ],
        allowedAttributes: {
          'a': [ 'href' ]
        },
        allowedIframeHostnames: []
    }).trim();

    const textStripped = sanitizeHtml(text, {
        allowedTags: [],
        allowedAttributes: {},
        allowedIframeHostnames: []
    }).trim();      

    return [textHtml, textStripped];
}

export const isOverflowed = (id: string):boolean => {
    const element = document.getElementById(id);
    
    if(element == null) {
        return false;
    }

    return element.offsetHeight < element.scrollHeight;
}

export const followUser = async (userId: string, followUserId: string, shouldFollow: boolean) => {
    if(userId === followUserId) {
        return true;
    }

    const data = {
        follow: shouldFollow,
        userId: userId,
        followId: followUserId
    };

    const result = await postSetFollowStatus(data);
    
    return result.status === 200;
}

export const getPfpFromProfile = (profile: Profile|null):string => {
    if(profile === null || profile.pfp == null || profile.pfp.length <= 0) {
        return DEFAULT_PFP;
    }

    return profile.pfp;
}

export const getPfpFromPost = (post: Post):string => {
    if(post.user.pfp == null || post.user.pfp.length === 0) {
       /* const result = postGetProfileByUserId({userId: post.user.userId});
        if(result == null || result.status !== 200) {
            return DEFAULT_PFP;
        }        
        return result.data.pfp;*/
        return DEFAULT_PFP;
    }    

    return post.user.pfp;
}

export const splitFullName = (fullName:string) => {
    const names = fullName.trim().split(' ');
    const firstName = names[0];
    const lastName = names[names.length - 1];
    const middleNames = names.slice(1, names.length - 1).join(' ');

    return { firstName, middleNames, lastName };
}
