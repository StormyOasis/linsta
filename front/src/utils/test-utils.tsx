import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RootState } from "../Components/Redux/redux";
import { authInitialState } from '../Components/Redux/slices/auth.slice';
import { modalInitialState } from '../Components/Redux/slices/modals.slice';
import { profileInitialState } from '../Components/Redux/slices/profile.slice';
import { miscInitialState } from '../Components/Redux/slices/misc.slice';
import { configureStore } from '@reduxjs/toolkit';

import Theme from '../Components/Themes/Theme';

export const mockInitialState = (overrides?: Partial<RootState>): RootState => ({
    auth: { ...authInitialState, ...(overrides?.auth ?? {}) },
    modal: { ...modalInitialState, ...(overrides?.modal ?? {}) },
    profile: { ...profileInitialState, ...(overrides?.profile ?? {}) },
    misc: { ...miscInitialState, ...(overrides?.misc ?? {}) },
});

type ExtendedRenderOptions = {
    store?: ReturnType<typeof configureStore>;
    preloadedState?: any;
    route?: string; // ‚Üê allow overriding the initial route
} & Omit<RenderOptions, "queries">;

const dummyAuthReducer = (state = { user: null, isLoading: false, error: null }, _action: any) => state;
const dummyModalReducer = (state = { isOpen: false }, _action: any) => state;
const dummyProfileReducer = (state = { name: '', email: '' }, _action: any) => state;
const dummyMiscReducer = (state = { someData: null }, _action: any) => state;

const renderWithStore = (
    ui: ReactElement,
    {
        preloadedState,
        store = configureStore({
            reducer: {
                auth: dummyAuthReducer,
                modal: dummyModalReducer,
                profile: dummyProfileReducer,
                misc: dummyMiscReducer,
            },
            preloadedState,
        }),
        route = "/",
        ...renderOptions
    }: ExtendedRenderOptions = {}
) => {
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
        return (
            <Provider store={store}>
                <MemoryRouter initialEntries={[route]}>
                    <Theme>{children}</Theme>
                </MemoryRouter>
            </Provider>
        );
    };

    return {
        ...render(ui, { wrapper: Wrapper, ...renderOptions }),
        store,
    };
};

export * from "@testing-library/react";
export { renderWithStore };
export type { ExtendedRenderOptions };