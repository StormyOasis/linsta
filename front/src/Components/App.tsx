import React, { useEffect, useState } from "react"
import { useLocation, useNavigate } from "react-router-dom";
import { ErrorBoundary } from "react-error-boundary";

import Layout from "../Components/Layout/Layout";
import Theme from "../Components/Themes/Theme";
import { historyUtils } from "../utils/utils";
import { useSelector } from "react-redux";
import { useAppDispatch } from "./Redux/redux";
import { getProfileByUserId } from "./Redux/slices/profile.slice";

const App: React.FC<any> = ({ children }) => {
    const [isServer, setIsServer] = useState<boolean>(true);
    const authUser = useSelector((value: any) => value?.auth?.user);

    const dispatch = useAppDispatch();

    //UseEffect only runs on the client      
    useEffect(() => {
        setIsServer(false);

        if (authUser && authUser.id != null) {
            dispatch(getProfileByUserId({ userId: authUser.id }));
        }
    }, [authUser]);

    historyUtils.location = useLocation();
    historyUtils.navigate = useNavigate();
    historyUtils.isServer = isServer;

    return (
        <React.StrictMode>
            <ErrorBoundary FallbackComponent={({ error }) => <div data-testid="fallback">{error.stack}</div>}>
                {children}
                <Theme>
                    <Layout />
                </Theme>
            </ErrorBoundary>
        </React.StrictMode>
    )
}

export default App;