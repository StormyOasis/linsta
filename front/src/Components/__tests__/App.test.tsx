import React from 'react'
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';
import App from '../App';
import { Provider } from 'react-redux';
import { buildStore } from '../Redux/redux';
import { historyUtils } from '../../utils/utils';


describe("Primary App Component", () => {
    it("should render logged out correctly", () => {
        const store = buildStore();
        const r = render(<MemoryRouter><Provider store={store}><App /></Provider></MemoryRouter>).asFragment();
        expect(r).toMatchSnapshot();
    });

    it("should render logged in correctly", () => {
        historyUtils.isServer = true;
        const store = buildStore({auth: {user: {id: 19, token: "abcde"}}});        
        console.log(store.getState());
        const r = render(<MemoryRouter><Provider store={store}><App /></Provider></MemoryRouter>).asFragment();
        expect(r).toMatchSnapshot();
    });    
})
