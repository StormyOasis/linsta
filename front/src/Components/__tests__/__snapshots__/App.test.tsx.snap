// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primary App Component should render logged in correctly 1`] = `
<DocumentFragment>
  <div
    class="Layout__ModalContainer-sc-1y0f571-1 dljdPV"
    id="modalContainer"
  />
  <section
    class="Layout__Section-sc-1y0f571-0 kMCVf"
    id="mainSectionContainer"
  >
    <main
      class="MainLayout__MainWrapper-sc-1gsojru-0 bzvUxv"
      role="main"
    >
      MainLayout
    </main>
  </section>
</DocumentFragment>
`;

exports[`Primary App Component should render logged out correctly 1`] = `
<DocumentFragment>
  <div
    class="Layout__ModalContainer-sc-1y0f571-1 dljdPV"
    id="modalContainer"
  />
  <section
    class="Layout__Section-sc-1y0f571-0 kMCVf"
    id="mainSectionContainer"
  >
    <main
      class="LoginLayout__LoginLayoutWrapper-sc-1p2723n-0 ilDFLd"
      role="main"
    >
      <div>
        <div>
          <div>
            <div
              class="LargeLogo__LogoWrapper-sc-1bu4w53-0 hBpxrf"
            >
              <a
                href="/"
              />
            </div>
            <form
              class="LoginLayout__LoginForm-sc-1p2723n-1 iKZqXv"
              method="post"
            >
              <div>
                <span>
                  Log in to see photos and videos from your friends.
                </span>
              </div>
              <div
                class="StyledInput__StyledInputWrapper-sc-fkf055-0 hOmQjA"
              >
                <label
                  class="StyledInput__StyledInputLabel-sc-fkf055-2 gizmBW"
                >
                  <input
                    class="StyledInput__StyledInputInput-sc-fkf055-1 ejdHNL"
                    maxlength="255"
                    name="userName"
                    placeholder="Username"
                    value=""
                  />
                </label>
              </div>
              <div
                class="StyledInput__StyledInputWrapper-sc-fkf055-0 hOmQjA"
              >
                <label
                  class="StyledInput__StyledInputLabel-sc-fkf055-2 gizmBW"
                >
                  <input
                    class="StyledInput__StyledInputInput-sc-fkf055-1 ejdHNL"
                    maxlength="255"
                    name="password"
                    placeholder="Password"
                    type="password"
                    value=""
                  />
                </label>
              </div>
              <button
                class="StyledButton__StyledButtonComponent-sc-8t4kac-0 fLkIsR"
                disabled=""
                style="margin-top: 8px;"
                type="submit"
              >
                Login
              </button>
              <div
                style="width: 100%;"
              >
                <div
                  style="margin-top: -5px;"
                >
                  <div>
                    <div>
                      <div>
                        <div />
                        <div>
                          OR
                        </div>
                        <div />
                      </div>
                    </div>
                    <button
                      class="LoginWithFB__LoginWithFBButton-sc-18oes83-0 eSywRS"
                    >
                      <span />
                      Log in with Facebook
                    </button>
                  </div>
                </div>
              </div>
              <div>
                <a
                  class="StyledLink__StyledLinkWrapper-sc-1rgr01q-0 dLJqDG"
                  href="/forgot"
                >
                  Forgot Password?
                </a>
              </div>
            </form>
          </div>
          <div
            style="padding: 20px 0px;"
          >
            Don't have an account? 
            <a
              class="StyledLink__StyledLinkWrapper-sc-1rgr01q-0 dLJqDG"
              href="/signup"
            >
              Sign up
            </a>
          </div>
        </div>
      </div>
    </main>
  </section>
</DocumentFragment>
`;
